### PR TITLE
feat: clickbait shield

### PR DESCRIPTION
**Pull request by omBratteng**
## Changes

Adds clickbait shield to cards and posts when we have detected that the post has a clickbait title.

## Events

Did you introduce any new tracking events?

## Experiment

Did you introduce any new experiments?

## Manual Testing

> [!CAUTION]
> Please make sure existing components are not breaking/affected by this PR

AS-788

### Preview domain
https://as-788-clickbait-shield.preview.app.daily.dev

### CODE REVIEW

**Review by rebelchris — APPROVED** [review]
Looks good to me, let's wait for one more high level review.
Please ping people if not responsive :)

File: packages/shared/src/components/modals/ClickbaitShieldModal.tsx [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  fetchSmartTitle,
+  ...props
+}: Props & ModalProps): ReactElement => {
+  const { logSubscriptionEvent } = usePlusSubscription();
+  const { closeModal } = useLazyModal();
+  const isMobile = useViewSize(ViewSize.MobileL);
+  if (!isMobile) {
+    closeModal();
+    return null;
+  }
```
Comment (capJavert) [inline packages/shared/src/components/modals/ClickbaitShieldModal.tsx:35]: Why is this needed?

File: packages/shared/src/components/plus/PlusList.tsx [hunk 2]
```diff
@@ -60,7 +60,7 @@ export const plusFeatureList: Array<PlusItem> = [
   },
   {
     label: 'AI-powered clickbait-free titles',
-    status: PlusItemStatus.ComingSoon,
+    status: PlusItemStatus.Ready,
```
Comment (capJavert) [inline packages/shared/src/components/plus/PlusList.tsx:63]: Nice ✅

File: packages/shared/src/hooks/post/useSmartTitle.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    ...getPostByIdKey(post?.id),
+  );
+
+  const { data: title, refetch } = useQuery({
+    queryKey: key,
+    queryFn: async () => {
+      // Enusre that we don't accidentally fetch the smart title for users outside of the feature flag
+      if (!showPlusSubscription || !isLoggedIn) {
+        return post?.title || post?.sharedPost?.title;
+      }
```
Comment (capJavert) [inline packages/shared/src/hooks/post/useSmartTitle.ts:42]: `showPlusSubscription` should be part of the query key so if user gets enrolled for any reason the query refetches automatically.

File: packages/shared/src/hooks/post/useSmartTitle.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    staleTime: Infinity,
+    ...disabledRefetch,
+  });
+
+  const fetchSmartTitle = useCallback(async () => {
+    if (!fetchedSmartTitle) {
+      await refetch();
+    } else {
+      client.setQueryData(key, post?.title);
+    }
```
Comment (capJavert) [inline packages/shared/src/hooks/post/useSmartTitle.ts:75]: Is this the case where we toggle back so we just reuse default title?

File: packages/shared/src/hooks/usePostById.ts [hunk 5]
```diff
@@ -129,7 +131,9 @@ const usePostById = ({ id, options = {} }: UsePostByIdProps): UsePostById => {
     staleTime: StaleTime.Default,
     enabled: !!id && tokenRefreshed,
   });
-  const post = postById || (options?.initialData as PostData);
+  const post = isPlus
+    ? postById
+    : postById || (options?.initialData as PostData);
```
Comment (capJavert) [inline packages/shared/src/hooks/usePostById.ts:136]: Think we should revert this https://github.com/dailydotdev/apps/pull/3930/files#r1872952334

**Review by capJavert — COMMENTED** [review]
Some questions, but don't see anything too blocking except for SSG revert. Nice work 🙌

File: packages/shared/src/components/modals/ClickbaitShieldModal.tsx (hunk 1 above)
Comment (omBratteng) [reply packages/shared/src/components/modals/ClickbaitShieldModal.tsx:35]: Because we don't want to show the modal on other devices

File: packages/shared/src/hooks/post/useSmartTitle.ts (hunk 4 above)
Comment (omBratteng) [reply packages/shared/src/hooks/post/useSmartTitle.ts:75]: This is the case when we want to look at the other title

File: packages/shared/src/hooks/post/useSmartTitle.ts (hunk 3 above)
Comment (omBratteng) [reply packages/shared/src/hooks/post/useSmartTitle.ts:49]: Added it as part of the key

File: packages/shared/src/hooks/post/useSmartTitle.ts (hunk 4 above)
Comment (capJavert) [reply packages/shared/src/hooks/post/useSmartTitle.ts:82]: yeah thats what I meant, toggling back and forth

File: packages/shared/src/hooks/post/useSmartTitle.ts [hunk 6]
```diff
@@ -26,8 +26,15 @@ export const useSmartTitle = (post: Post): UseSmartTitle => {
...
 
-  const key = useMemo(() => [...getPostByIdKey(post?.id), 'title'], [post?.id]);
-  const trialKey = generateQueryKey(
+  const key = useMemo(
+    () => [
+      ...getPostByIdKey(post?.id),
+      'title',
+      showPlusSubscription.toString(),
+    ],
+    [post?.id, showPlusSubscription],
```
Comment (capJavert) [inline packages/shared/src/hooks/post/useSmartTitle.ts:35]: We can maybe do ` {key: 'title', showPlusSubscription: true},` just so query key does not grow indefinitely

File: packages/shared/src/hooks/usePostById.ts (hunk 5 above)
Comment (omBratteng) [reply packages/shared/src/hooks/usePostById.ts:136]: The problem is that when using the SSG, its value gets set as initial data in the useSmartTitle query state, so you'll see the original title even if you have clickbait shield enabled

File: packages/shared/src/hooks/post/useSmartTitle.ts (hunk 6 above)
Comment (omBratteng) [reply packages/shared/src/hooks/post/useSmartTitle.ts:35]: Not sure what you mean?

File: packages/shared/src/hooks/usePostById.ts (hunk 5 above)
Comment (capJavert) [reply packages/shared/src/hooks/usePostById.ts:136]: Maybe avoid using initialData, initialData is more like a default in useState, so its tricky because it only takes the first ever value, here we need to conditionally set one or the other

Comment (omBratteng) [reply packages/shared/src/hooks/usePostById.ts:136]: I think I got it to work, opening separate PR for it to test

Comment (omBratteng) [reply packages/shared/src/hooks/usePostById.ts]: Reverted it in #3932

**Review by capJavert — APPROVED** [review]
